test/dhcp/free_ips_test.rb

```diff
@@ -0,0 +1,108 @@
+require 'test_helper'
+require 'set'
+require 'dhcp_common/dhcp_common'
+require 'dhcp_common/subnet'
+require 'dhcp_common/record/reservation'
+require 'dhcp_common/free_ips'
+
+class Proxy::DHCPFreeIpsTest < Test::Unit::TestCase
+  def setup
+    @free_ips = Proxy::DHCP::FreeIps.new
+    @subnet = Proxy::DHCP::Subnet.new("192.168.1.0", "255.255.255.0")
+  end
+
+  def test_random_index
+    indices = Set.new
+    @free_ips.random_index(6) {|i| indices << i}
+    assert_equal 6, indices.size
+    assert_equal Set.new([0, 1, 2, 3, 4, 5]), indices
+  end
+
+  def test_random_index_for_array_of_size_one
+    indices = Set.new
+    @free_ips.random_index(1) {|i| indices << i}
+    assert_equal Set.new([0]), indices
+  end
+
+  def test_address_range_with_start_and_end
+    assert_equal [3_232_235_620, 100], @free_ips.address_range_with_start_and_end("192.168.0.100", "192.168.0.200")
+  end
+
+  def test_find_free_ip
+    @free_ips.stubs(:icmp_pingable?).returns(false)
+    @free_ips.stubs(:tcp_pingable?).returns(false)
+    r = Proxy::DHCP::Reservation.new('test', "192.168.1.1", "aa:bb:cc:dd:ee:ff", @subnet, :hostname =>'test')
+
+    assert_equal "192.168.1.2", @free_ips.find_free_ip("192.168.1.1", "192.168.1.2", [r])
+  end
+
+  def test_find_free_ip_should_return_nil_when_no_addresses_available
+    @free_ips.stubs(:icmp_pingable?).returns(false)
+    @free_ips.stubs(:tcp_pingable?).returns(false)
+    r = Proxy::DHCP::Reservation.new('test', "192.168.1.1", "aa:bb:cc:dd:ee:ff", @subnet, :hostname =>'test')
+
+    assert_nil @free_ips.find_free_ip("192.168.1.1", "192.168.1.1", [r])
+  end
```

Comment:
I'd also add a edge case when all addresses are blacklisted, just to make sure. Nitpick: no space in `hostname => 'test'`.